os/drivers/seclink/seclink_drv.c

```diff
@@ -0,0 +1,202 @@
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static const struct file_operations g_seclink_fops = {
+	seclink_open,                                                   /* open */
+	seclink_close,                                                   /* close */
+	seclink_read,
+	seclink_write,
+	0,                                                   /* seek */
+	seclink_ioctl                                        /* ioctl */
+#ifndef CONFIG_DISABLE_POLL
+	, 0                                                  /* poll */
+#endif
+};
+
+int seclink_open(FAR struct file *filep)
+{
+	SLDRV_ENTER;
+
+	FAR struct inode *inode = filep->f_inode;
+	FAR struct sec_upperhalf_s *upper = inode->i_private;
+	upper->refcnt++;
+
+	return 0;
+}
+
+int seclink_close(FAR struct file *filep)
+{
+	SLDRV_ENTER;
+
+	FAR struct inode *inode = filep->f_inode;
+	FAR struct sec_upperhalf_s *upper = inode->i_private;
+	upper->refcnt--;
+
+	return 0;
+}
+
+ssize_t seclink_read(FAR struct file *filep, FAR char *buffer, size_t len)
+{
+	SLDRV_ENTER;
+
+	return 0;
+}
+
+ssize_t seclink_write(FAR struct file *filep, FAR const char *buffer, size_t len)
+{
+	SLDRV_ENTER;
+
+	return 0;
+}
+
+int seclink_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
+{
+	SLDRV_ENTER;
+
+	SLDRV_LOG("-->%s (%d)(%x)\n", __FUNCTION__, cmd, arg);
+
+	FAR struct inode *inode = filep->f_inode;
+	FAR struct sec_upperhalf_s *upper = inode->i_private;
+
+	/* If there are a couple of tasks that request SE then
+	 * calling lower's function should be protected. But
+	 * current design consider that only one task that is seclink can
+	 * request to HAL
+	 */
+	if (cmd == SECLINK_HAL_INIT || cmd == SECLINK_HAL_DEINIT) {
+		hd_handle_common_request(cmd, arg, (void *)upper->lower);
+	} else if (cmd & SECLINK_AUTH) {
+		hd_handle_auth_reqeust(cmd, arg, (void *)upper->lower);
+	} else if (cmd & SECLINK_KEY) {
+		hd_handle_key_request(cmd, arg, (void *)upper->lower);
+	} else if (cmd & SECLINK_SS) {
+		hd_handle_ss_request(cmd, arg, (void *)upper->lower);
+	} else if (cmd & SECLINK_CRYPTO) {
+		hd_handle_crypto_request(cmd, arg, (void *)upper->lower);
+	}
+
+	return 0;
+}
+
+/****************************************************************************
+ * Name: se_register
+ *
+ * Description:
+ *   Register Secure Element device.
+ *
+ ****************************************************************************/
+int se_register(const char *path, struct sec_lowerhalf_s *lower)
+{
+	vdbg("Registering %s\n", path);
+
+	if (!lower) {
+		return -1;
+	}
+
+	struct sec_upperhalf_s *upper = (struct sec_upperhalf_s *)malloc(sizeof(struct sec_upperhalf_s));
+	if (!upper) {
+		return -1;
+	}
+
+	/*  initialize upper */
+	char *drv_path = (char *)malloc(strlen(path) + 1);
+	if (!drv_path) {
+		free(upper);
+		return -1;
+	}
+	memcpy(drv_path, path, strlen(path) + 1);
```

Comment:
"drv_path" is used nowhere in the below. Is it correct..?

Reply:
Thanks I'll fix